lib/Sema/CSDiagnostics.cpp

```diff
@@ -1355,6 +1355,22 @@ bool MemberAccessOnOptionalBaseFailure::diagnoseAsError() {
           .fixItInsert(sourceRange.End, "!.");
     }
   } else {
+    // Check whether or not the base of this optional unwrap is implicit self
+    //  - This can only happen with a [weak self] capture, and is not permitted.
+    if (auto dotExpr = getAsExpr<UnresolvedDotExpr>(locator->getAnchor())) {
```

Comment:
I still don't think this is the correct strategy. We shouldn't rely on the shape of the AST coming into and out of pre-check to make this call.

Reply:
What do you mean? This is checking conditions for emitting a tailored diagnostic for cases when the base expression of this member access is optional implicit `self`. 

I'm not sure what you mean by "we shouldn't rely on the shape of the AST coming into and out of pre-check". Is there a better way to check if this invalid member access is on an implicit `self` dealt?

Reply:
@CodaFi, could you share additional input on this?